src/cnxml/de.ts

```diff
@@ -0,0 +1,679 @@
+const FIGURE = {
+    caption,
+    media,
+    subfigure: block('figure', { media, caption }),
+}
+
+/** Block elements */
+const BLOCK = {
+    ...LINE,
+    definition,
+    exercise: block('exercise', EXERCISE),
+    figure: block('figure', FIGURE),
+    note,
+    rule,
+}
+
+/** Content of most mixed elements */
+const MIXED = { ...LINE, ...INLINE }
+
+/** Media items */
+const MEDIA = { audio: mediaItem, image: mediaItem, video: mediaItem }
+
+/** Contents of a rule */
+const RULE = {
+    example: block('rule_example', LINE),
+    proof: block('rule_proof', LINE),
+    statement: block('rule_statement', LINE),
+    title,
+}
+
+/** Contents of a list */
+const LIST = { item }
+
+/** Contents of a glossary */
+const GLOSSARY = { definition }
+
+/** Deserializators to use when one couldn't be found in context */
+const FALLBACK: Deserializers = { ...INLINE, ...BLOCK, ...MEDIA }
+
+/* --- deserializers -------------------------------------------------------- */
+
+/** Deserialize <code> */
+function code(editor: Editor, el: Element, at: Path) {
+    buildElement(editor, el, at, {
+        type: 'code',
+        language: el.getAttribute('lang'),
+        placement: el.getAttribute('display') === 'block' ? 'block' : 'line',
+    }, INLINE)
+    // NOTE: this element contains pre-formatted content and should not be
+    // normalized.
+}
+
+/** Deserialize <content> */
+function content(editor: Editor, el: Element) {
+    children(editor, el, [0], BLOCK)
+    normalizeBlock(editor, [])
+}
+
+/** Deserialize a <term> in <definition> */
+function definitionTerm(editor: Editor, el: Element, at: Path) {
+    buildElement(editor, el, at, {
+        type: 'definition_term',
+        reference: el.getAttributeNS(CMLNLE_NAMESPACE, 'reference'),
+    }, INLINE)
+    normalizeLine(editor, at)
+}
+
+/** Deserialize <foreign> */
+function foreign(editor: Editor, el: Element, at: Path) {
+    buildElement(editor, el, at, {
+        type: 'foreign',
+        language: el.getAttributeNS(XML_NAMESPACE, 'lang') ,
+    }, INLINE)
+    normalizeLine(editor, at)
+}
+
+/** Deserialize <glossary> */
+function glossary(editor: Editor, el: Element) {
+    const path = [editor.children.length]
+    buildElement(editor, el, path, { type: 'glossary' }, GLOSSARY)
+    normalizeBlock(editor, path)
+}
+
+/** Deserialize a list item */
+function item(editor: Editor, el: Element, at: Path) {
+    buildElement(editor, el, at, { type: 'list_item' }, MIXED)
+    normalizeMixed(editor, at)
+
+    const node = Slate.Node.get(editor, at)
+    if (node.children.length === 1 && List.isList(node.children[0])) {
+        Transforms.unwrapNodes(editor, { at })
+    }
+}
+
+/** Deserialize a cross-reference, a document reference, or a hyperlink */
+function link(editor: Editor, el: Element, at: Path) {
+    const target = el.getAttribute('target-id')
+    const document = el.getAttribute('document')
+    const url = el.getAttribute('url')
+
+    if (target != null) {
```

Comment:
Shouldn't check for `url` be above this one?

Reply:
When `url` is used `document`, `version`, `target-id`, and `resource` must not be present, so in a correct document order of these test doesn't matter.